crates/wasmtime/src/instance.rs

```diff
@@ -183,10 +191,27 @@ impl Instance {
             bail!("cross-`Engine` instantiation is not currently supported");
         }
 
-        let info = module.register_frame_info();
-        store.register_jit_code(module.compiled_module().jit_code_ranges());
+        let host_info = Box::new({
+            let frame_info_registration = module.register_frame_info();
+            store.register_jit_code(module.compiled_module().jit_code_ranges());
 
-        let handle = instantiate(store, module.compiled_module(), imports, Box::new(info))?;
+            // We need to make sure that we keep this alive as long as the instance
+            // is alive, or else we could miss GC roots, reclaim objects too early,
+            // and get user-after-frees.
+            let stack_map_registration =
+                unsafe { module.register_stack_maps(&*store.stack_map_registry()) };
+
+            (frame_info_registration, stack_map_registration)
+        });
+
+        let handle = instantiate(
+            store,
+            module.compiled_module(),
+            imports,
+            host_info,
+            store.externref_activations_table().clone(),
+            store.stack_map_registry().clone(),
```

Comment:
you can probably avoid passing these two parameters explicitly and read them from the passed-in `store` argument inside the method.